cpp/arcticdb/column_store/column.hpp

```diff
@@ -693,74 +716,175 @@ class Column {
     }
 
     template<
-        typename left_input_tdt,
-        typename right_input_tdt,
-        typename output_tdt,
-        typename functor,
-        typename = std::enable_if_t<std::is_invocable_r_v<
+            typename left_input_tdt,
+            typename right_input_tdt,
+            typename output_tdt,
+            typename functor>
+    requires std::is_invocable_r_v<
             typename output_tdt::DataTypeTag::raw_type,
             functor,
             typename left_input_tdt::DataTypeTag::raw_type,
-            typename right_input_tdt::DataTypeTag::raw_type>>>
+            typename right_input_tdt::DataTypeTag::raw_type>
     static void transform(const Column& left_input_column,
                           const Column& right_input_column,
                           Column& output_column,
                           functor&& f) {
         auto left_input_data = left_input_column.data();
         auto right_input_data = right_input_column.data();
+        initialise_output_column(left_input_column, right_input_column, output_column);
         auto output_data = output_column.data();
-        std::transform(left_input_data.cbegin<left_input_tdt>(), left_input_data.cend<left_input_tdt>(), right_input_data.cbegin<right_input_tdt>(), output_data.begin<output_tdt>(), std::forward<functor>(f));
+        auto output_it = output_data.begin<output_tdt>();
+
+        if (!left_input_column.is_sparse() && !right_input_column.is_sparse()) {
+            // Both dense, use std::transform over the shorter column to avoid going out-of-bounds
+            if (left_input_column.row_count() <= right_input_column.row_count()) {
+                std::transform(left_input_data.cbegin<left_input_tdt>(),
+                               left_input_data.cend<left_input_tdt>(),
+                               right_input_data.cbegin<right_input_tdt>(),
+                               output_it,
+                               std::forward<functor>(f));
+            } else {
+                std::transform(right_input_data.cbegin<left_input_tdt>(),
+                               right_input_data.cend<left_input_tdt>(),
+                               left_input_data.cbegin<right_input_tdt>(),
+                               output_it,
+                               std::forward<functor>(f));
+            }
+        } else if (left_input_column.is_sparse() && right_input_column.is_sparse()) {
+            // Both sparse, only project rows in the intersection of on-bits from both sparse maps
+            auto left_accessor = random_accessor<left_input_tdt>(&left_input_data);
+            auto right_accessor = random_accessor<right_input_tdt>(&right_input_data);
+            // TODO: experiment with more efficient bitset traversal methods
+            // https://github.com/tlk00/BitMagic/tree/master/samples/bvsample25
+            for (auto set_bit = output_column.sparse_map().first(); set_bit < output_column.sparse_map().end(); ++set_bit) {
```

Comment:
Do we know how expensive calculating sparse_map().end() is? It might be safer to hoist it up out of the loop control